standard_rules/unusual_login.py

```diff
@@ -19,62 +23,80 @@ def rule(event):
-    url = "https://ipinfo.io/" + event.udm("source_ip") + "/geo"
-
-    # Skip API call if this is a unit test
-    if "panther_api_data" in event:
-        resp = lambda: None
-        setattr(resp, "status_code", 200)
-        setattr(resp, "text", event.get("panther_api_data"))
-    else:
-        # This response looks like the following:
-        # {‘ip': '8.8.8.8', 'city': 'Mountain View', 'region': 'California', 'country': 'US',
-        # 'loc': '37.4056,-122.0775', 'postal': '94043', 'timezone': 'America/Los_Angeles'}
-        resp = requests.get(url)
-
-    if resp.status_code != 200:
-        raise Exception(f"API call failed: GET {url} returned {resp.status_code}")
-    login_info = json.loads(resp.text)
-    # The idea is to create a fingerprint of this login, and then keep track of all the fingerprints
-    # for a given user's logins. In this way, we can detect unusual logins.
-    login_tuple = login_info.get("region", "<REGION>") + ":" + login_info.get("city", "<CITY>")
-    EVENT_LOGIN_INFO[event.get("p_row_id")] = login_tuple
-
-    # Lookup & store persistent data
+    # Mocked during unit testing
+    GEO_INFO = geoinfo_from_ip(event.udm("source_ip"))
+
+    # As of Panther 1.19, mocking returns all mocked objects in a string
+    # GEO_INFO must be converted back to a dict to mimic the API call
+    if isinstance(GEO_INFO, str):
+        GEO_INFO = json.loads(GEO_INFO)
+
+    # Look up history of unique geolocations
     event_key = get_key(event)
-    last_login_info = get_string_set(event_key)
-    fingerprint_timestamp = str(datetime.datetime.now())
-    if not last_login_info:
-        # Store this as the first login if we've never seen this user login before
-        put_string_set(event_key, [json.dumps({login_tuple: fingerprint_timestamp})])
+    # Mocked during unit testing
+    previous_geo_logins = get_string_set(event_key)
+
+    # As of Panther 1.19, mocking returns all mocked objects in a string
+    # previous_geo_logins must be converted back to a set to mimic the API call
+    if isinstance(previous_geo_logins, str):
+        print("previous_geo_logins is a mocked string:")
+        print(previous_geo_logins)
+        previous_geo_logins = ast.literal_eval(previous_geo_logins)
+        print("new type of previous_geo_logins is:", type(previous_geo_logins))
+
+    new_login_geo = (
+        f"{GEO_INFO.get('region', '<UNKNOWN_REGION>')}"
+        ":"
+        f"{GEO_INFO.get('city', '<UNKNOWN_CITY>')}"
+    )
+    new_login_timestamp = event.get("p_event_time", "")
+
+    # convert set of single string to dictionary
+    previous_geo_logins = json.loads(previous_geo_logins.pop())
+
+    # don't alert if the geo is already in the history
+    if previous_geo_logins.get(new_login_geo):
+        # update timestamp of the existing geo in the history
+        previous_geo_logins[new_login_geo] = new_login_timestamp
+        # exclude Dynamo API call from unit test
+        if "mock" not in event:
```

Comment:
rather than rely on adding `mock` to the event, could we add a yml mock definition for `put_string_set`?

Reply:
Yes, I think we should remove this conditional and just mock `put_string_set` to return `""`.